drv/transceivers-server/src/udp.rs

```diff
@@ -149,64 +154,57 @@ impl ServerImpl {
-                // anything with this message, and the protocol mandates that we
-                // drop the message.
-                if header.version < version::MIN {
-                    ringbuf_entry!(Trace::WrongVersion(header.version));
-                    None
-
-                // In this case, the message has failed to deserialize, but
-                // we've successfully deserialized the header. That implies that
-                // the host has sent us a message that does not exist in our own
-                // implementation of the protocol.
-                //
-                // Check that implication by ensuring that the host version is
-                // after our own CURRENT.
-                } else if header.version > version::CURRENT {
-                    let n = hubpack::serialize(
-                        tx_data_buf,
-                        &Message {
-                            header,
-                            modules: ModuleId::default(),
-                            body: MessageBody::SpResponse(SpResponse::Error(
-                                Error::VersionMismatch {
-                                    expected: version::CURRENT,
-                                    actual: header.version,
-                                },
-                            )),
-                        },
-                    );
-                    Some(n.unwrap() as u32)
-
-                // This last case catches failures to deserialize something we
-                // _should_ have been able to deserialize. The host version is
-                // between MIN and CURRENT, so it's a message that's part of our
-                // own protocol. We really just failed to deserialize it, such
-                // as a corrupt buffer or some similar failure mode. Just drop
-                // the message, since we've already logged a `DeserializeError`.
-                } else {
-                    None
-                }
-            }
-            Err(e) => {
-                ringbuf_entry!(Trace::DeserializeHeaderError(e));
-                None
-            }
+        // This message comes from a host implementation older than the
+        // minimum committed version of the protocol. We really can't do
+        // anything with this message, and the protocol mandates that we
+        // drop the message.
+        if header.version() < version::outer::MIN {
+            ringbuf_entry!(Trace::WrongVersion(header.version()));
+            None
+
+        // In this case, the message has failed to deserialize, but
+        // we've successfully deserialized the header. That implies that
+        // the host has sent us a message that does not exist in our own
+        // implementation of the protocol.
+        //
+        // Check that implication by ensuring that the host version is
+        // after our own CURRENT.
+        } else if header.version() > version::outer::CURRENT {
```

Comment:
I think this captures some of my ambivalence about the inner / outer protocol version split.

What we really want to check here is the _inner_ protocol version. The logic is: we've deserialized the header but not the message, implying that the message is too new for us. But we can't really check that, since we can't deserialize the inner version. So this is only the correct check if the inner and outer versions are the same, which limits the value of the split.

I think we have two options here:

- Make a single version
- Try to deserialize the _version alone_, i.e., use the knowledge about the inner format of the `Message` type to read the first octet of the remaining bytes of the message, which should have the version.

I'm really not sure which to take. Making the inner / outer versions the same would be the easiest, but then we'd probably want to go back to a single message type that contains both the header and the body. That's also fine, it's just more churn that I think we're both hoping to avoid.